### PR TITLE
Metrics "Sidecar" (scraping container) support

### LINKED ISSUE #2986: Support metrics API
dashboard should read from the metrics APIs in order to be compatible with future kubernetes versions.

heapster is being deprecated (https://github.com/kubernetes/heapster/pull/2022) in favor of metrics-server (https://github.com/kubernetes-incubator/metrics-server).

The current heapster or metrics-server releases can provide the metrics APIs (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/resource-metrics-api.md). So a single client implementation can support both on I think kubernetes 1.7 or above.

cc #2976, #1310

### PR DESCRIPTION
**Pull request by jeefy**
**WIP**

This is the Dashboard side for using the metrics server. Fixes #2986 

- [x] Functional Support (through the additional container)
- [x] Finish rebasing to "v2.0.0-alpha"
- [x] Official Repo for the scraping container (now at https://github.com/kubernetes-sigs/dashboard-metrics-scraper)
- [x] Docker Hub dev registry for the scraping container
- [x] Update `head` deploy scripts to include scraping container

### CODE REVIEW

**Comment by jeefy** [comment]
I'm not going to block on the gcr.io repo at this point. 

At this point I'm good moving this into `master`, tackling any emergent bugs, and getting people testing both. 

To test, people will have to use `aio/deploy/recommended/kubernetes-dashboard-head.yaml` -- However, `head` will test both `head` of Dashboard, as well as `head` of metrics-scraper.

:tada: 

/hold remove
/assign @maciaszczykm @floreks

File: src/app/backend/dashboard.go [hunk 1]
```diff
@@ -116,8 +121,19 @@ func main() {
 
 	// Init integrations
 	integrationManager := integration.NewIntegrationManager(clientManager)
-	integrationManager.Metric().ConfigureHeapster(args.Holder.GetHeapsterHost()).
-		EnableWithRetry(integrationapi.HeapsterIntegrationID, time.Duration(args.Holder.GetMetricClientCheckPeriod()))
+
+	fmt.Println("Hi dude")
```
Comment (dorant) [inline src/app/backend/dashboard.go]: Just some drive-by comments: old printouts?

**Comment by jeefy** [comment]
-_- Thank you for that. I'll remove that and see if I left any other
artifacts.

**Comment by gregory-lyons** [comment]
@jeefy any status update on this PR?

**Comment by jeefy** [comment]
This should be good to merge once it gets reviewed. I'll look to push for more reviewers once I'm done traveling this week. 

In the meantime I welcome anyone's feedback.

**Comment by jeefy** [comment]
@maciaszczykm @floreks PTAL -- It's ready for review but may require some improvements once more eyes are on it.

File: src/app/backend/integration/metric/sidecar/restclient.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+// InClusterSidecarClient is an in-cluster implementation of a Sidecar client. Talks with Sidecar
+// through service proxy.
+type inClusterSidecarClient struct {
+	client rest.Interface
+}
+
+// Get creates request to given path.
+func (c inClusterSidecarClient) Get(path string) RequestInterface {
+	return c.client.Get().
+		Namespace("kube-system").
```
Comment (danbeaulieu) [inline src/app/backend/integration/metric/sidecar/restclient.go]: Will the sidecar service always live in the kube-system namespace?

File: src/app/backend/integration/metric/manager.go [hunk 3]
```diff
@@ -40,7 +41,9 @@ type MetricManager interface {
 	EnableWithRetry(id integrationapi.IntegrationID, period time.Duration)
 	// List returns list of available metric related integrations.
 	List() []integrationapi.Integration
-	// ConfigureHeapster configures and adds heapster to clients list.
+	// ConfigureSidecar configures and adds sidecar to clients list.
+	ConfigureSidecar(host string) MetricManager
+	// ConfigureSidecar configures and adds sidecar to clients list.
```
Comment (danbeaulieu) [inline src/app/backend/integration/metric/manager.go]: I think you want this comment to be for the ConfigureHeapster function.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml [hunk 4]
```diff
@@ -12,31 +12,6 @@
...
 # limitations under the License.
 
-# ------------------- Dashboard Secrets ------------------- #
-
-apiVersion: v1
-kind: Secret
-metadata:
-  labels:
-    k8s-app: kubernetes-dashboard-head
-  name: kubernetes-dashboard-certs
```
Comment (floreks) [inline aio/deploy/recommended/kubernetes-dashboard-head.yaml]: I think it shouldn't be removed.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml [hunk 5]
```diff
@@ -12,31 +12,6 @@
 # See the License for the specific language governing permissions and
```
Comment (floreks) [inline aio/deploy/recommended/kubernetes-dashboard-head.yaml]: Looks like this file was rebased to some old version.

File: src/app/backend/dashboard.go [hunk 6]
```diff
@@ -116,8 +121,15 @@ func main() {
 
 	// Init integrations
 	integrationManager := integration.NewIntegrationManager(clientManager)
-	integrationManager.Metric().ConfigureHeapster(args.Holder.GetHeapsterHost()).
-		EnableWithRetry(integrationapi.HeapsterIntegrationID, time.Duration(args.Holder.GetMetricClientCheckPeriod()))
+
+	if args.Holder.GetMetricsProvider() == "sidecar" {
```
Comment (floreks) [inline src/app/backend/dashboard.go]: This could be a switch since only 1 metric provider can be active anyway

File: src/app/backend/integration/metric/manager.go (hunk 3 above)
Comment (floreks) [inline src/app/backend/integration/metric/manager.go]: s/ConfigureSidecar/ConfigureHeapster

File: src/app/backend/integration/metric/sidecar/restclient.go (hunk 2 above)
Comment (floreks) [reply src/app/backend/integration/metric/sidecar/restclient.go]: I think it should live in the same namespace as Dashboard. You can use `args.Holder.getNamespace()` instead.

File: src/app/backend/integration/metric/sidecar/restclient.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		Name("sidecar").
+		SubResource("proxy").
+		Suffix(path)
+}
+
+// HealthCheck does a health check of the application.
+// Returns nil if connection to application can be established, error object otherwise.
+func (self inClusterSidecarClient) HealthCheck() error {
+	_, err := self.client.Get().
+		Namespace("kube-system").
```
Comment (floreks) [inline src/app/backend/integration/metric/sidecar/restclient.go]: Same here.

File: src/app/backend/integration/api/types.go [hunk 8]
```diff
@@ -24,6 +24,7 @@ type IntegrationID string
 // Integration app IDs should be registered in this block.
 const (
 	HeapsterIntegrationID IntegrationID = "heapster"
```
Comment (maciaszczykm) [inline src/app/backend/integration/api/types.go]: Do we plan to remove Heapster soon?

File: src/app/backend/integration/metric/sidecar/client.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// Sidecar client implements MetricClient and Integration interfaces.
+type sidecarClient struct {
+	client SidecarRESTClient
+}
+
+// Implement Integration interface.
```
Comment (maciaszczykm) [inline src/app/backend/integration/metric/sidecar/client.go]: TODO?

File: src/app/backend/integration/metric/sidecar/client.go [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+	return self.client.HealthCheck()
+}
+
+// ID implements integration app interface. See Integration interface for more information.
+func (self sidecarClient) ID() integrationapi.IntegrationID {
+	return integrationapi.SidecarIntegrationID
+}
+
+// Implement MetricClient interface
```
Comment (maciaszczykm) [inline src/app/backend/integration/metric/sidecar/client.go]: TODO?

File: src/app/backend/integration/metric/manager.go (hunk 3 above)
Comment (jeefy) [reply src/app/backend/integration/metric/manager.go]: You are correct. :) Thanks!

File: src/app/backend/integration/api/types.go (hunk 8 above)
Comment (jeefy) [reply src/app/backend/integration/api/types.go]: I was thinking that we'd have one or two releases that supported both Heapster and Metrics Server, then remove Heapster support entirely.

If consensus is to rip the bandaid off and remove Heapster support with this PR, I can do that pretty easily. Just let me know. :)

File: src/app/backend/integration/metric/sidecar/client.go (hunk 9 above)
Comment (jeefy) [reply src/app/backend/integration/metric/sidecar/client.go]: No, the Sidecar code borrows heavily (read: copied, pasted, then modified) from the Heapster metric integration. The Heapster integration has the [same comment](https://github.com/kubernetes/dashboard/blob/master/src/app/backend/integration/metric/heapster/client.go#L39) in it.

File: src/app/backend/integration/metric/sidecar/client.go (hunk 10 above)
Comment (jeefy) [reply src/app/backend/integration/metric/sidecar/client.go]: Same as above.

File: src/app/backend/integration/api/types.go (hunk 8 above)
Comment (maciaszczykm) [reply src/app/backend/integration/api/types.go]: @floreks WDYT?

File: src/app/backend/integration/metric/sidecar/restclient.go (hunk 2 above)
Comment (jeefy) [reply src/app/backend/integration/metric/sidecar/restclient.go]: Fixed

File: src/app/backend/integration/metric/manager.go (hunk 3 above)
Comment (jeefy) [reply src/app/backend/integration/metric/manager.go]: Fixed. :)

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml (hunk 4 above)
Comment (jeefy) [reply aio/deploy/recommended/kubernetes-dashboard-head.yaml]: You are correct. :) Fixed.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml (hunk 5 above)
Comment (jeefy) [reply aio/deploy/recommended/kubernetes-dashboard-head.yaml]: Yeah, not sure from where. :\ PTAL now.

File: src/app/backend/dashboard.go (hunk 6 above)
Comment (jeefy) [reply src/app/backend/dashboard.go]: Done!

File: src/app/backend/integration/metric/manager.go (hunk 3 above)
Comment (jeefy) [reply src/app/backend/integration/metric/manager.go]: Derp. Fixed. :)

**Comment by brandond** [comment]
@jeefy do you have a public image for this that I can test? I see that you have something up at jeefy/kubernetes-dashboard-amd64 on docker-hub but those appear to be out of date.

File: src/app/backend/dashboard.go [hunk 11]
```diff
@@ -56,10 +56,15 @@ var (
 		"to connect to in the format of protocol://address:port, e.g., "+
 		"http://localhost:8080. If not specified, the assumption is that the binary runs inside a "+
 		"Kubernetes cluster and local discovery is attempted.")
-	argHeapsterHost = pflag.String("heapster-host", "", "The address of the Heapster Apiserver "+
+	argMetricsProvider = pflag.String("metrics-provider", "sidecar", "Select provider type for metrics. Defaults to 'sidecar'")
```
Comment (brandond) [inline src/app/backend/dashboard.go]: Don't include the default in the help text, this is already printed by the help system:
```
      --metrics-provider string          Select provider type for metrics. Defaults to 'sidecar' (default "sidecar")
```
Consider instead listing the options: "sidecar" or "heapster".

Comment (jeefy) [reply src/app/backend/dashboard.go]: I'm picking this comment to say: You're awesome and I'm going to try to go through all these in the next couple days. Thanks so much!

**Comment by TDeprez** [comment]
Maybe it's just the wording, but would this also work with something like kube-prometheus (prometheus operator)? It uses prometheus to provide metrics for the APIService, removing the need to use a metrics server in the kubernetes cluster.

**Comment by jeefy** [comment]
This will not work with Prometheus. This is meant to be a shim, as Heapster was removed in favor of the Metrics Server. 

We have plans to support Prometheus this year, but it's not part of this effort. :(

**Comment by brandond** [comment]
@jeefy would it make sense to have the sidecar also serve prometheus metrics, or should that go directly into metrics-server?

**Comment by jeefy** [comment]
It's an interesting question/approach. 

I _think_ the sidecar should only focus on one thing (metrics-server shim) and then Prometheus become natively supported within Dashboard. 

I also think that adding Prometheus support into the metrics-sidecar would be easier and faster than getting it to work in the Dashboard.

I personally don't want to make a decision on that till A) More people have discussed and B) The current PR gets merged in.

File: src/app/backend/integration/api/types.go (hunk 8 above)
Comment (floreks) [reply src/app/backend/integration/api/types.go]: I'd do as @jeefy suggests. Leave for a couple of releases until heapster is completely removed.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml [hunk 12]
```diff
@@ -121,9 +145,8 @@ spec:
     spec:
       containers:
       - name: kubernetes-dashboard-head
-        image: kubernetesdashboarddev/kubernetes-dashboard-amd64:head
-        # Image is tagged and updated with :head, so always pull it.
-        imagePullPolicy: Always
+        image: kubernetes/kubernetes-dashboard-amd64:head
+        imagePullPolicy: Never
```
Comment (floreks) [inline aio/deploy/recommended/kubernetes-dashboard-head.yaml]: As head frequently changes, I'd leave pull policy to always.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml [hunk 13]
```diff
@@ -121,9 +145,8 @@ spec:
     spec:
       containers:
       - name: kubernetes-dashboard-head
-        image: kubernetesdashboarddev/kubernetes-dashboard-amd64:head
-        # Image is tagged and updated with :head, so always pull it.
-        imagePullPolicy: Always
+        image: kubernetes/kubernetes-dashboard-amd64:head
```
Comment (floreks) [inline aio/deploy/recommended/kubernetes-dashboard-head.yaml]: Does it exist?

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml [hunk 14]
```diff
@@ -69,40 +69,64 @@ rules:
   resources: ["secrets"]
   resourceNames: ["kubernetes-dashboard-key-holder", "kubernetes-dashboard-certs", "kubernetes-dashboard-csrf"]
   verbs: ["get", "update", "delete"]
-  # Allow Dashboard to get and update 'kubernetes-dashboard-settings' config map.
+  # Allow Dashboard to get and update 'kubernetes-dashboard-head-settings' config map.
```
Comment (floreks) [inline aio/deploy/recommended/kubernetes-dashboard-head.yaml]: The resource name below is still `kubernetes-dashboard-settings`. Same for key holder.

File: src/app/backend/dashboard.go [hunk 15]
```diff
@@ -116,8 +121,17 @@ func main() {
...
+
+	switch metricsProvider := args.Holder.GetMetricsProvider(); metricsProvider {
+	case "sidecar":
+		integrationManager.Metric().ConfigureSidecar(args.Holder.GetSidecarHost()).
+			EnableWithRetry(integrationapi.SidecarIntegrationID, time.Duration(args.Holder.GetMetricClientCheckPeriod()))
+	case "heapster":
+		integrationManager.Metric().ConfigureHeapster(args.Holder.GetHeapsterHost()).
+			EnableWithRetry(integrationapi.HeapsterIntegrationID, time.Duration(args.Holder.GetMetricClientCheckPeriod()))
+	default:
+		log.Printf("Invalid or metrics provider selected: %s", metricsProvider)
```
Comment (floreks) [inline src/app/backend/dashboard.go]: I'd do something like `Invalid metrics provider selected ... Defaulting to 'sidecar'`.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml (hunk 13 above)
Comment (jeefy) [reply aio/deploy/recommended/kubernetes-dashboard-head.yaml]: Mmmmmm nope. Bad copy pasta.

File: aio/deploy/recommended/kubernetes-dashboard-head.yaml (hunk 14 above)
Comment (jeefy) [reply aio/deploy/recommended/kubernetes-dashboard-head.yaml]: Bad copy pasta/edit. Thanks.

File: src/app/backend/dashboard.go (hunk 15 above)
Comment (jeefy) [reply src/app/backend/dashboard.go]: Good idea. :+1:

**Comment by jseguillon** [comment]
I'm very interested by this feature 👍 is there any way I could help ?

**Comment by shu-mutou** [comment]
@jeefy I'm trying to test this PR, but I got following error and backend seems to stuck. Could you advice for this?
```
[0] 2019/06/03 06:06:35 [2019-06-03T06:06:35Z] Incoming HTTP/1.1 GET /api/v1/node?itemsPerPage=10&page=1&sortBy=d,creationTimestamp request from 127.0.0.1:35984: {}
[0] panic: runtime error: index out of range
[0] 
[0] goroutine 457 [running]:
github.com/kubernetes/dashboard/src/app/backend/integration/metric/sidecar.sidecarClient.ithResourceDownload.func1(0x1685420, 0xc0005010b0, 0x1465a8d, 0x4, 0x14669c3, 0x6, 0xc00039a460, 0x1, 0x1, 0xc0008fa450, ...)
[0] 	/go/src/github.com/kubernetes/dashboard/.tmp/backend/src/github.com/kubernetes/dashboard/src/app/backend/integration/metric/sidecar/client.go:163 +0x4c1
[0] created by github.com/kubernetes/dashboard/src/app/backend/integration/metric/sidecar.sidecarClient.ithResourceDownload
[0] 	/go/src/github.com/kubernetes/dashboard/.tmp/backend/src/github.com/kubernetes/dashboard/src/app/backend/integration/metric/sidecar/client.go:154 +0x11e
```

**Comment by maciaszczykm** [comment]
@jeefy We have made progress in the router migration and the release should happen this month. Is it ready for the final review?

**Comment by jeefy** [comment]
@maciaszczykm @floreks I think it's time for a review.

In this PR I also added a "local" deployment YAML. For reference, here's how I just tested it.

```
minikube start
minikube addons list # ensure metrics-server addon is enabled
eval $(minikube docker-env)
npm run docker:build:head
kubectl apply -f aio/deploy/recommended/kubernetes-dashboard-local.yaml
kubectl proxy
```

And then viewing the dashboard via kubectl proxy. Let me know if there's anything else needed to dive into this.

File: src/app/backend/dashboard.go (hunk 11 above)
Comment (maciaszczykm) [reply src/app/backend/dashboard.go]: Will you change that?

File: aio/deploy/recommended/kubernetes-dashboard-local.yaml [hunk 16]
```diff
@@ -0,0 +1,257 @@
+# Copyright 2017 The Kubernetes Authors.
```
Comment (floreks) [inline aio/deploy/recommended/kubernetes-dashboard-local.yaml]: What is this file `kubernetes-dashboard-local`?

File: src/app/backend/dashboard.go (hunk 15 above)
Comment (floreks) [reply src/app/backend/dashboard.go]: `Invalid or metrics provider selected` is it correct?

**Comment by floreks** [comment]
I wouldn't add this local deployment file to our `aio/deploy` directory as it should contain official deployments. From my understanding, this deployment will only work if we first build an image locally and then try to run it.

If we want to keep it then I'd move it maybe to a `test-resources` dir, rename it to something like `metrics-test-deployment.yaml` and also rename dashboard deployment inside the file so it does not collide with the official head deployment.

@jeefy @maciaszczykm WDYT?

**Comment by maciaszczykm** [comment]
@floreks SGTM.

**Comment by jeefy** [comment]
Sounds good, I'll update the PR later this morning.

**Comment by pires** [comment]
@jeefy I'm here rooting for you. What a champ you have been with this PR 🙇

**Comment by PierluigiLenociAkelius** [comment]
@ianlewis @PeWu please take a look 😘

**Comment by maciaszczykm** [comment]
LGTM, cc @floreks.

**Comment by floreks** [comment]
Let's get this in! :champagne: 

/lgtm

**Comment by jeefy** [comment]
Thanks for the support y'all!

image

**Comment by eloyekunle** [comment]
Thanks @jeefy

**Comment by jseguillon** [comment]
Wow @jeefy you're the boss ! Thanks !

**Comment by RainingNight** [comment]
Good!

**Comment by tdmalone** [comment]
Thanks so much for this @jeefy! Just to clarify, once a new version is cut (although I note it has been last year since this happened...), this means heapster won't be required anymore and Dashboard will be able to display metrics from metrics-server?

**Comment by jeefy** [comment]
@tdmalone Correct. Dashboard will (now) deploy two containers, the dashboard itself and the [Dashboard Metrics Scraper](https://github.com/kubernetes-sigs/dashboard-metrics-scraper). Since metrics-server only shows a single moment of time in the cluster, the scraper is needed to poll/store the metrics for look-up. The next thing to tackle is being able to point the Dashboard at Prometheus, but that'll be easier to do after the efforts of our GSOC students (Yay!)

Also, we should be cutting an RC at the end of this month, but you can (technically) spin up the Dashboard off the master branch using the [kubernetes-dashboard-head.yaml](https://raw.githubusercontent.com/kubernetes/dashboard/master/aio/deploy/recommended/kubernetes-dashboard-head.yaml) deployment. 

Just keep in mind, that's literally a build off master, not an actual release, there be dragons there, etc etc. 😄

File: src/app/backend/dashboard.go (hunk 11 above)
Comment (jeefy) [reply src/app/backend/dashboard.go]: I totally missed this, and I'll PR in a change this week.

**Comment by varuzam** [comment]
I run the kubernetes dashboard with the manifest https://gist.githubusercontent.com/varuzam/d624547b4286dd11206e1fc0a442f78d/raw/579905246348002b7b9895c60b3ae996de3f664a/kube-dashboard.yaml
The dashboard can not get metrics from the scrapper. Here is logs of the scrapper:
```
{"level":"info","msg":"Database updated: 2 nodes, 7 pods","time":"2019-06-18T19:05:06Z"}
{"level":"info","msg":"URL: /","time":"2019-06-18T19:05:14Z"}
{"level":"info","msg":"URL: /","time":"2019-06-18T19:05:24Z"}
{"level":"info","msg":"URL: **/api/v1/api/v1/**dashboard/namespaces/kube-system/pod-list/kubernetes-dashboard-86df9bb89b-l2226,metrics-server-54fddf7759-t8nnr,coredns-58b4966dd4-5gjxx,coredns-58b4966dd4-p4cn6/metrics/cpu/usage_rate","time":"2019-06-18T19:05:30Z"}
```
It looks like the dashboard uses wrong url with doubled '/api/v1'

**Comment by floreks** [comment]
@varuzam how did you run the scraper? In or outside of the cluster? It does work when both Dashboard and the scraper are running inside. There is an issue if I am trying to access it using remote option.

**Comment by varuzam** [comment]
@floreks I run the scraper in the one pod with the dashboard inside a cluster.
My deployment - https://gist.githubusercontent.com/varuzam/d624547b4286dd11206e1fc0a442f78d/raw/579905246348002b7b9895c60b3ae996de3f664a/kube-dashboard.yaml

**Comment by BouweCeunen** [comment]
Downloading logs from the dashboard gives me a 404, I deployed it with the kubernetes-dashboard-head.yaml @jeefy provided.
